File: gfhttpc-restructured/gfarm-http.py
#!/usr/bin/env python3
"""
gfarm-http - HTTP client for Gfarm filesystem operations
"""

import os
import sys
import json
import argparse
import urllib.parse
import urllib.request
import urllib.error
import base64
import ssl
from pathlib import Path


def get_jwt_token():
    """Get JWT token from file or environment"""
    jwt_path = os.environ.get('JWT_USER_PATH')
    if not jwt_path:
        uid = os.getuid()
        jwt_path = f"/tmp/jwt_user_u{uid}/token.jwt"

    if os.path.exists(jwt_path):
        with open(jwt_path, 'r') as f:
            return f.read().strip()
    return None


def get_auth_headers():
    """Get authentication headers for requests"""
    sasl_user = os.environ.get('GFARM_SASL_USER')
    sasl_password = os.environ.get('GFARM_SASL_PASSWORD')

    if sasl_user == "anonymous":
        return {}
    elif sasl_user and sasl_password:
        auth_str = base64.b64encode(
            f"{sasl_user}:{sasl_password}".encode()).decode()
        return {"Authorization": f"Basic {auth_str}"}
    else:
        token = get_jwt_token()
        if token:
            return {"Authorization": f"Bearer {token}"}
        else:
            print("Error: Environment variable JWT_USER_PATH (or, "
                  "GFARM_SASL_USER and GFARM_SASL_PASSWORD) is required",
                  file=sys.stderr)
            sys.exit(1)


def make_http_request(url, method="GET", data=None, headers=None,
                      output_file=None, upload_file=None, verbose=False,
                      insecure=False):
    """Make authenticated HTTP request using urllib"""
    auth_headers = get_auth_headers()

    # Prepare headers
    req_headers = {}
    req_headers.update(auth_headers)
    if headers:
        req_headers.update(headers)

    # Prepare data and content type
    request_data = None
    if data:
        if isinstance(data, dict):
            request_data = json.dumps(data).encode('utf-8')
            req_headers['Content-Type'] = 'application/json'
        else:
            request_data = (data.encode('utf-8')
                            if isinstance(data, str) else data)

    # Handle file upload
    if upload_file:
        if upload_file == "-":
            request_data = sys.stdin.buffer.read()
        else:
            upload_path = Path(upload_file)
            if upload_path.exists():
                # Add file timestamp header
                mtime = int(upload_path.stat().st_mtime)
                req_headers['X-File-Timestamp'] = str(mtime)
                with open(upload_path, 'rb') as f:
                    request_data = f.read()
            else:
                print(f"Error: File not found: {upload_file}", file=sys.stderr)
                sys.exit(1)

    # Create request
    req = urllib.request.Request(url, data=request_data,
                                 headers=req_headers, method=method)

    # Create SSL context
    ssl_context = ssl.create_default_context()
    if insecure:
        ssl_context.check_hostname = False
        ssl_context.verify_mode = ssl.CERT_NONE

    try:
        if verbose:
            print(f"Making {method} request to {url}", file=sys.stderr)
            for key, value in req_headers.items():
                print(f"Header: {key}: {value}", file=sys.stderr)

        with urllib.request.urlopen(req, context=ssl_context) as response:
            response_data = response.read()

            # Handle output
            if output_file:
                if output_file == "-":
                    sys.stdout.buffer.write(response_data)
                else:
                    with open(output_file, 'wb') as f:
                        f.write(response_data)
                    # Set file timestamp if available
                    if 'Last-Modified' in response.headers:
                        from email.utils import parsedate_to_datetime
                        try:
                            dt = parsedate_to_datetime(
                                response.headers['Last-Modified'])
                            timestamp = dt.timestamp()
                            os.utime(output_file, (timestamp, timestamp))
                        except Exception:
                            pass
            else:
                # Print to stdout
                try:
                    print(response_data.decode('utf-8'), end='')
                except UnicodeDecodeError:
                    sys.stdout.buffer.write(response_data)

    except urllib.error.HTTPError as e:
        error_data = e.read()
        try:
            error_text = error_data.decode('utf-8')
            print(f"HTTP {e.code} Error: {error_text}", file=sys.stderr)
        except UnicodeDecodeError:
            print(f"HTTP {e.code} Error", file=sys.stderr)
            sys.stderr.buffer.write(error_data)
        sys.exit(1)
    except urllib.error.URLError as e:
        print(f"URL Error: {e.reason}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error making HTTP request: {e}", file=sys.stderr)
        sys.exit(1)


def url_path_encode(path):
    """URL encode path while preserving forward slashes"""
    return urllib.parse.quote(path.lstrip('/'), safe='/')


def get_url_base():
    """Get the base URL from environment"""
    url = os.environ.get('GFARM_HTTP_URL')
    if not url:
        print("Error: GFARM_HTTP_URL is required", file=sys.stderr)
        sys.exit(1)
    return url.rstrip('/')


def get_common_options(args):
    """Get common HTTP options from args"""
    return {
        'verbose': getattr(args, 'verbose', False),
        'insecure': getattr(args, 'insecure', False)
    }


def cmd_ls(args):
    """List directory contents"""
    if not args.path:
        print("Error: Gfarm-path is required", file=sys.stderr)
        sys.exit(1)

    url_base = get_url_base()
    gfarm_path = url_path_encode(args.path[0])

    params = []
    if args.all:
        params.append("show_hidden=1")
    if args.effective:
        params.append("effperm=1")
    if args.long:
        params.append("long_format=1")
    if args.recursive:
        params.append("recursive=1")

    params_str = "?" + "&".join(params) if params else ""
    url = f"{url_base}/dir/{gfarm_path}{params_str}"

    opts = get_common_options(args)
    make_http_request(url, **opts)


def cmd_download(args):
    """Download file from Gfarm"""
    if len(args.path) != 2:
        print("Error: Both Gfarm-path and Local-path are required",
              file=sys.stderr)
        sys.exit(1)

    url_base = get_url_base()
    gfarm_path = url_path_encode(args.path[0])
    local_path = args.path[1]

    url = f"{url_base}/file/{gfarm_path}"

    opts = get_common_options(args)
    make_http_request(url, output_file=local_path, **opts)


def cmd_upload(args):
    """Upload file to Gfarm"""
    if len(args.path) != 2:
        print("Error: Both Local-path and Gfarm-path are required",
              file=sys.stderr)
        sys.exit(1)

    url_base = get_url_base()
    local_path = args.path[0]
    gfarm_path = url_path_encode(args.path[1])

    url = f"{url_base}/file/{gfarm_path}"

    opts = get_common_options(args)
    make_http_request(url, method="PUT", upload_file=local_path, **opts)


def cmd_mkdir(args):
    """Create directory"""
    if not args.path:
        print("Error: Gfarm-path is required", file=sys.stderr)
        sys.exit(1)

    url_base = get_url_base()
    gfarm_path = url_path_encode(args.path[0])

    url = f"{url_base}/dir/{gfarm_path}"

    opts = get_common_options(args)
    make_http_request(url, method="PUT", **opts)


def cmd_rm(args):
    """Remove file"""
    if not args.path:
        print("Error: Gfarm-path is required", file=sys.stderr)
        sys.exit(1)

    url_base = get_url_base()
    gfarm_path = url_path_encode(args.path[0])

    url = f"{url_base}/file/{gfarm_path}"

    opts = get_common_options(args)
    make_http_request(url, method="DELETE", **opts)


def cmd_rmdir(args):
    """Remove directory"""
    if not args.path:
        print("Error: Gfarm-path is required", file=sys.stderr)
        sys.exit(1)

    url_base = get_url_base()
    gfarm_path = url_path_encode(args.path[0])

    url = f"{url_base}/dir/{gfarm_path}"

    opts = get_common_options(args)
    make_http_request(url, method="DELETE", **opts)


def cmd_chmod(args):
    """Change file permissions"""
    if len(args.path) != 2:
        print("Error: Both mode and Gfarm-path are required", file=sys.stderr)
        sys.exit(1)

    url_base = get_url_base()
    mode = args.path[0]
    gfarm_path = url_path_encode(args.path[1])

    url = f"{url_base}/attr/{gfarm_path}"

    opts = get_common_options(args)
    data = {"Mode": mode}
    make_http_request(url, method="POST", data=data, **opts)


def cmd_stat(args):
    """Get file status"""
    if not args.path:
        print("Error: Gfarm-path is required", file=sys.stderr)
        sys.exit(1)

    url_base = get_url_base()
    gfarm_path = url_path_encode(args.path[0])

    url = f"{url_base}/attr/{gfarm_path}"

    opts = get_common_options(args)
    make_http_request(url, **opts)


def cmd_mv(args):
    """Move/rename file"""
    if len(args.path) != 2:
        print("Error: Both source and destination paths are required",
              file=sys.stderr)
        sys.exit(1)

    url_base = get_url_base()
    src_path = url_path_encode(args.path[0])
    dst_path = args.path[1]

    url = f"{url_base}/file/{src_path}"

    opts = get_common_options(args)
    data = {"Destination": dst_path}
    make_http_request(url, method="PATCH", data=data, **opts)


def cmd_whoami(args):
    """Show current user"""
    url_base = get_url_base()
    url = f"{url_base}/conf/me"

    opts = get_common_options(args)
    make_http_request(url, **opts)


def main():
    parser = argparse.ArgumentParser(
        description="HTTP client for Gfarm filesystem operations",
        formatter_class=argparse.RawDescriptionHelpFormatter
    )

    # Common options
    parser.add_argument("-k", "--insecure", action="store_true",
                        help="Insecure connection")
    parser.add_argument("-v", "--verbose", action="store_true",
                        help="Verbose mode")

    subparsers = parser.add_subparsers(dest='command',
                                       help='Available commands')

    # ls command
    ls_parser = subparsers.add_parser('ls', help='List directory contents')
    ls_parser.add_argument('-a', '--all', action='store_true',
                           help="Do not hide entries starting with '.'")
    ls_parser.add_argument('-e', '--effective', action='store_true',
                           help="Display effective permissions")
    ls_parser.add_argument('-l', '--long', action='store_true',
                           help="List in long format")
    ls_parser.add_argument('-R', '--recursive', action='store_true',
                           help="Recursively list subdirectories")
    ls_parser.add_argument('path', nargs=1, help='Gfarm path')

    # download command
    dl_parser = subparsers.add_parser('download',
                                      help='Download file from Gfarm')
    dl_parser.add_argument('path', nargs=2, help='Gfarm-path Local-path')

    # upload command
    up_parser = subparsers.add_parser('upload', help='Upload file to Gfarm')
    up_parser.add_argument('path', nargs=2, help='Local-path Gfarm-path')

    # mkdir command
    mkdir_parser = subparsers.add_parser('mkdir', help='Create directory')
    mkdir_parser.add_argument('path', nargs=1, help='Gfarm path')

    # rm command
    rm_parser = subparsers.add_parser('rm', help='Remove file')
    rm_parser.add_argument('path', nargs=1, help='Gfarm path')

    # rmdir command
    rmdir_parser = subparsers.add_parser('rmdir', help='Remove directory')
    rmdir_parser.add_argument('path', nargs=1, help='Gfarm path')

    # chmod command
    chmod_parser = subparsers.add_parser('chmod',
                                         help='Change file permissions')
    chmod_parser.add_argument('path', nargs=2, help='mode(octal) Gfarm-path')

    # stat command
    stat_parser = subparsers.add_parser('stat', help='Get file status')
    stat_parser.add_argument('path', nargs=1, help='Gfarm path')

    # mv command
    mv_parser = subparsers.add_parser('mv', help='Move/rename file')
    mv_parser.add_argument('path', nargs=2, help='source destination')

    # whoami command
    subparsers.add_parser('whoami', help='Show current user')

    args = parser.parse_args()

    if not args.command:
        parser.print_help()
        sys.exit(1)

    # Command dispatch
    commands = {
        'ls': cmd_ls,
        'download': cmd_download,
        'upload': cmd_upload,
        'mkdir': cmd_mkdir,
        'rm': cmd_rm,
        'rmdir': cmd_rmdir,
        'chmod': cmd_chmod,
        'stat': cmd_stat,
        'mv': cmd_mv,
        'whoami': cmd_whoami,
    }

    if args.command in commands:
        commands[args.command](args)
    else:
        print(f"Error: Unknown command '{args.command}'", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: gfhttpc-restructured/gfarm-http.go
package main

import (
	"bytes"
	"context"
	"crypto/tls"
	"encoding/base64"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const (
	DefaultDirMode  = 0o755
	DefaultFileMode = 0o644
	DefaultTimeout  = 30 * time.Second
)

type Config struct {
	BaseURL    string
	Verbose    bool
	Insecure   bool
	HTTPClient *http.Client
}

type HTTPError struct {
	StatusCode int
	Message    string
}

func (e *HTTPError) Error() string {
	return fmt.Sprintf("HTTP %d: %s", e.StatusCode, e.Message)
}

func (c *Config) GetAuthHeaders() (map[string]string, error) {
	h := make(map[string]string)

	user := os.Getenv("GFARM_SASL_USER")
	pass := os.Getenv("GFARM_SASL_PASSWORD")

	if user == "anonymous" {
		return h, nil
	}
	if user != "" && pass != "" {
		auth := base64.StdEncoding.EncodeToString([]byte(user + ":" + pass))
		h["Authorization"] = "Basic " + auth
		return h, nil
	}

	tok, err := getJWTToken()
	if err != nil {
		return nil, fmt.Errorf("read JWT token: %w", err)
	}
	if tok != "" {
		h["Authorization"] = "Bearer " + tok
		return h, nil
	}
	return nil, fmt.Errorf("environment GFARM_SASL_USER/GFARM_SASL_PASSWORD or JWT_USER_PATH is required")
}

func getJWTToken() (string, error) {
	jwtPath := os.Getenv("JWT_USER_PATH")
	if jwtPath == "" {
		jwtPath = fmt.Sprintf("/tmp/jwt_user_u%d/token.jwt", os.Getuid())
	}
	b, err := os.ReadFile(jwtPath)
	if errors.Is(err, os.ErrNotExist) {
		return "", nil
	}
	if err != nil {
		return "", err
	}
	tok := strings.TrimSpace(string(b))
	if tok == "" {
		return "", fmt.Errorf("JWT token file is empty: %s", jwtPath)
	}
	return tok, nil
}

// Encode each path segment but keep slashes as separators.
func encodePath(p string) string {
	p = strings.TrimPrefix(p, "/")
	segs := strings.Split(p, "/")
	for i, s := range segs {
		segs[i] = url.PathEscape(s)
	}
	return strings.Join(segs, "/")
}

func formatByteData(b []byte) string {
	var v any
	if json.Unmarshal(b, &v) == nil {
		out, _ := json.MarshalIndent(v, "", "  ")
		return string(out)
	}
	return string(b)
}

func (c *Config) vlogf(format string, a ...any) {
	if c.Verbose {
		fmt.Fprintf(os.Stderr, format, a...)
	}
}

func (c *Config) prepareRequestBody(data any, uploadFile string, headers map[string]string) (io.Reader, error) {
	if uploadFile != "" {
		return c.prepareUploadBody(uploadFile, headers)
	}
	if data != nil {
		return c.prepareJSONBody(data, headers)
	}
	return nil, nil
}

func (c *Config) prepareUploadBody(uploadFile string, headers map[string]string) (io.Reader, error) {
	if uploadFile == "-" {
		return os.Stdin, nil
	}

	f, err := os.Open(uploadFile)
	if err != nil {
		return nil, fmt.Errorf("open %s: %w", uploadFile, err)
	}
	// Note: caller must close the file
	if st, err := f.Stat(); err == nil {
		headers["X-File-Timestamp"] = strconv.FormatInt(st.ModTime().Unix(), 10)
	}
	return f, nil
}

func (c *Config) prepareJSONBody(data any, headers map[string]string) (io.Reader, error) {
	raw, err := json.Marshal(data)
	if err != nil {
		return nil, fmt.Errorf("marshal json: %w", err)
	}
	headers["Content-Type"] = "application/json"
	return bytes.NewReader(raw), nil
}

func (c *Config) createRequest(ctx context.Context, method, requestURL string, body io.Reader) (*http.Request, error) {
	req, err := http.NewRequestWithContext(ctx, method, requestURL, body)
	if err != nil {
		return nil, fmt.Errorf("new request: %w", err)
	}

	authHeaders, err := c.GetAuthHeaders()
	if err != nil {
		return nil, err
	}

	for k, v := range authHeaders {
		req.Header.Set(k, v)
	}
	return req, nil
}

func (c *Config) logRequest(req *http.Request) {
	if !c.Verbose {
		return
	}
	c.vlogf("=> %s %s\n", req.Method, req.URL.String())
	for k, vs := range req.Header {
		for _, v := range vs {
			c.vlogf("   %s: %s\n", k, v)
		}
	}
}

func (c *Config) handleResponse(resp *http.Response) ([]byte, error) {
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		eb, _ := io.ReadAll(resp.Body)
		return nil, &HTTPError{
			StatusCode: resp.StatusCode,
			Message:    formatByteData(eb),
		}
	}

	b, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response: %w", err)
	}
	return b, nil
}

func (c *Config) writeOutput(data []byte, outputFile string, resp *http.Response) error {
	if outputFile == "" {
		fmt.Print(formatByteData(data))
		return nil
	}

	if outputFile == "-" {
		_, err := os.Stdout.Write(data)
		return err
	}

	// Ensure parent directory exists
	if d := filepath.Dir(outputFile); d != "" && d != "." {
		if err := os.MkdirAll(d, DefaultDirMode); err != nil {
			return fmt.Errorf("create directory %s: %w", d, err)
		}
	}

	if err := os.WriteFile(outputFile, data, DefaultFileMode); err != nil {
		return fmt.Errorf("write %s: %w", outputFile, err)
	}

	// Set file timestamp if provided
	if lm := resp.Header.Get("Last-Modified"); lm != "" {
		if t, err := time.Parse(time.RFC1123, lm); err == nil {
			_ = os.Chtimes(outputFile, t, t)
		}
	}
	return nil
}

func (c *Config) makeHTTPRequest(method, requestURL string, data any, headers map[string]string, outputFile string, uploadFile string) error {
	ctx, cancel := context.WithTimeout(context.Background(), DefaultTimeout)
	defer cancel()

	if headers == nil {
		headers = make(map[string]string)
	}

	body, err := c.prepareRequestBody(data, uploadFile, headers)
	if err != nil {
		return err
	}

	// Close file if it was opened for upload
	if uploadFile != "" && uploadFile != "-" {
		if closer, ok := body.(io.Closer); ok {
			defer closer.Close()
		}
	}

	req, err := c.createRequest(ctx, method, requestURL, body)
	if err != nil {
		return err
	}

	for k, v := range headers {
		req.Header.Set(k, v)
	}

	c.logRequest(req)

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}

	responseData, err := c.handleResponse(resp)
	if err != nil {
		return err
	}

	return c.writeOutput(responseData, outputFile, resp)
}

// Commands
func (c *Config) cmdLs(path string, all, effective, longf, timef, recursive, json bool) error {
	if path == "" {
		return fmt.Errorf("gfarm-path is required")
	}
	gfarmPath := encodePath(path)
	params := url.Values{}
	if all {
		params.Set("show_hidden", "on")
	}
	if effective {
		params.Set("effperm", "on")
	}
	if longf {
		params.Set("long_format", "on")
	}
	if timef {
		params.Set("time_format", "full")
	} else {
		params.Set("time_format", "short")
	}
	if recursive {
		params.Set("recursive", "on")
	}
	if json {
		params.Set("output_format", "json")
	} else {
		params.Set("output_format", "plain")
	}
	u := fmt.Sprintf("%s/dir/%s", c.BaseURL, gfarmPath)
	if len(params) > 0 {
		u += "?" + params.Encode()
	}
	return c.makeHTTPRequest("GET", u, nil, nil, "", "")
}

func (c *Config) cmdDownload(gfarmPath, localPath string) error {
	if gfarmPath == "" || localPath == "" {
		return fmt.Errorf("both Gfarm-path and Local-path are required")
	}
	p := encodePath(gfarmPath)
	u := fmt.Sprintf("%s/file/%s", c.BaseURL, p)
	return c.makeHTTPRequest("GET", u, nil, nil, localPath, "")
}

func (c *Config) cmdUpload(localPath, gfarmPath string) error {
	if localPath == "" || gfarmPath == "" {
		return fmt.Errorf("both Local-path and Gfarm-path are required")
	}
	p := encodePath(gfarmPath)
	u := fmt.Sprintf("%s/file/%s", c.BaseURL, p)
	return c.makeHTTPRequest("PUT", u, nil, nil, "", localPath)
}

func (c *Config) cmdMkdir(path string) error {
	if path == "" {
		return fmt.Errorf("gfarm-path is required")
	}
	p := encodePath(path)
	u := fmt.Sprintf("%s/dir/%s", c.BaseURL, p)
	return c.makeHTTPRequest("PUT", u, nil, nil, "", "")
}

func (c *Config) cmdRm(path string) error {
	if path == "" {
		return fmt.Errorf("gfarm-path is required")
	}
	p := encodePath(path)
	u := fmt.Sprintf("%s/file/%s", c.BaseURL, p)
	return c.makeHTTPRequest("DELETE", u, nil, nil, "", "")
}

func (c *Config) cmdRmdir(path string) error {
	if path == "" {
		return fmt.Errorf("gfarm-path is required")
	}
	p := encodePath(path)
	u := fmt.Sprintf("%s/dir/%s", c.BaseURL, p)
	return c.makeHTTPRequest("DELETE", u, nil, nil, "", "")
}

func (c *Config) cmdChmod(mode, path string) error {
	if mode == "" || path == "" {
		return fmt.Errorf("both mode and Gfarm-path are required")
	}
	p := encodePath(path)
	u := fmt.Sprintf("%s/attr/%s", c.BaseURL, p)
	data := map[string]string{"Mode": mode}
	return c.makeHTTPRequest("POST", u, data, nil, "", "")
}

func (c *Config) cmdStat(path string) error {
	if path == "" {
		return fmt.Errorf("gfarm-path is required")
	}
	p := encodePath(path)
	u := fmt.Sprintf("%s/attr/%s", c.BaseURL, p)
	return c.makeHTTPRequest("GET", u, nil, nil, "", "")
}

func (c *Config) cmdMv(srcPath, dstPath string) error {
	if srcPath == "" || dstPath == "" {
		return fmt.Errorf("both source and destination paths are required")
	}
	p := encodePath(srcPath)
	u := fmt.Sprintf("%s/file/%s", c.BaseURL, p)
	data := map[string]string{"Destination": dstPath}
	return c.makeHTTPRequest("PATCH", u, data, nil, "", "")
}

func (c *Config) cmdWhoami() error {
	u := fmt.Sprintf("%s/conf/me", c.BaseURL)
	return c.makeHTTPRequest("GET", u, nil, nil, "", "")
}

// ---- CLI  ----

type Command struct {
	Name        string
	Description string
	Handler     func(*Config, []string) error
}

var commands = map[string]*Command{
	"ls": {
		Name:        "ls",
		Description: "List directory",
		Handler:     handleLs,
	},
	"download": {
		Name:        "download",
		Description: "Download file",
		Handler:     handleDownload,
	},
	"upload": {
		Name:        "upload",
		Description: "Upload file",
		Handler:     handleUpload,
	},
	"mkdir": {
		Name:        "mkdir",
		Description: "Create directory",
		Handler:     handleMkdir,
	},
	"rm": {
		Name:        "rm",
		Description: "Remove file",
		Handler:     handleRm,
	},
	"rmdir": {
		Name:        "rmdir",
		Description: "Remove directory",
		Handler:     handleRmdir,
	},
	"chmod": {
		Name:        "chmod",
		Description: "Change permissions",
		Handler:     handleChmod,
	},
	"stat": {
		Name:        "stat",
		Description: "File status",
		Handler:     handleStat,
	},
	"mv": {
		Name:        "mv",
		Description: "Move/rename",
		Handler:     handleMv,
	},
	"whoami": {
		Name:        "whoami",
		Description: "Show current user",
		Handler:     handleWhoami,
	},
}

func handleLs(cfg *Config, args []string) error {
	fs := flag.NewFlagSet("ls", flag.ExitOnError)
	all := fs.Bool("a", false, "do not hide entries starting with '.'")
	eff := fs.Bool("e", false, "display effective permissions")
	longf := fs.Bool("l", false, "list in long format")
	timef := fs.Bool("T", false, "With the -l option, show complete date format")
	rec := fs.Bool("R", false, "recursively list subdirectories")
	json := fs.Bool("j", false, "output with json format")

	if err := fs.Parse(args); err != nil {
		return err
	}
	rest := fs.Args()
	if len(rest) != 1 {
		return fmt.Errorf("ls requires exactly 1 Gfarm-path")
	}
	return cfg.cmdLs(rest[0], *all, *eff, *longf, *timef, *rec, *json)
}

func handleDownload(cfg *Config, args []string) error {
	fs := flag.NewFlagSet("download", flag.ExitOnError)
	if err := fs.Parse(args); err != nil {
		return err
	}
	rest := fs.Args()
	if len(rest) != 2 {
		return fmt.Errorf("download requires Gfarm-path and Local-path")
	}
	return cfg.cmdDownload(rest[0], rest[1])
}

func handleUpload(cfg *Config, args []string) error {
	fs := flag.NewFlagSet("upload", flag.ExitOnError)
	if err := fs.Parse(args); err != nil {
		return err
	}
	rest := fs.Args()
	if len(rest) != 2 {
		return fmt.Errorf("upload requires Local-path and Gfarm-path")
	}
	return cfg.cmdUpload(rest[0], rest[1])
}

func handleMkdir(cfg *Config, args []string) error {
	fs := flag.NewFlagSet("mkdir", flag.ExitOnError)
	if err := fs.Parse(args); err != nil {
		return err
	}
	rest := fs.Args()
	if len(rest) != 1 {
		return fmt.Errorf("mkdir requires exactly 1 Gfarm-path")
	}
	return cfg.cmdMkdir(rest[0])
}

func handleRm(cfg *Config, args []string) error {
	fs := flag.NewFlagSet("rm", flag.ExitOnError)
	if err := fs.Parse(args); err != nil {
		return err
	}
	rest := fs.Args()
	if len(rest) != 1 {
		return fmt.Errorf("rm requires exactly 1 Gfarm-path")
	}
	return cfg.cmdRm(rest[0])
}

func handleRmdir(cfg *Config, args []string) error {
	fs := flag.NewFlagSet("rmdir", flag.ExitOnError)
	if err := fs.Parse(args); err != nil {
		return err
	}
	rest := fs.Args()
	if len(rest) != 1 {
		return fmt.Errorf("rmdir requires exactly 1 Gfarm-path")
	}
	return cfg.cmdRmdir(rest[0])
}

func handleChmod(cfg *Config, args []string) error {
	fs := flag.NewFlagSet("chmod", flag.ExitOnError)
	if err := fs.Parse(args); err != nil {
		return err
	}
	rest := fs.Args()
	if len(rest) != 2 {
		return fmt.Errorf("chmod requires mode(octal) and Gfarm-path")
	}
	return cfg.cmdChmod(rest[0], rest[1])
}

func handleStat(cfg *Config, args []string) error {
	fs := flag.NewFlagSet("stat", flag.ExitOnError)
	if err := fs.Parse(args); err != nil {
		return err
	}
	rest := fs.Args()
	if len(rest) != 1 {
		return fmt.Errorf("stat requires exactly 1 Gfarm-path")
	}
	return cfg.cmdStat(rest[0])
}

func handleMv(cfg *Config, args []string) error {
	fs := flag.NewFlagSet("mv", flag.ExitOnError)
	if err := fs.Parse(args); err != nil {
		return err
	}
	rest := fs.Args()
	if len(rest) != 2 {
		return fmt.Errorf("mv requires source and destination")
	}
	return cfg.cmdMv(rest[0], rest[1])
}

func handleWhoami(cfg *Config, args []string) error {
	fs := flag.NewFlagSet("whoami", flag.ExitOnError)
	if err := fs.Parse(args); err != nil {
		return err
	}
	if len(fs.Args()) != 0 {
		return fmt.Errorf("whoami takes no arguments")
	}
	return cfg.cmdWhoami()
}

func exitErr(err error) {
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}

func showRootHelp() {
	fmt.Printf(`Usage: %s [global-options] <command> [command-options] [args]

Global options:
  -k, --insecure    Insecure TLS (skip verify)
  -v, --verbose     Verbose logging
  -h, --help        Show this help

Commands:
`, os.Args[0])
	for _, cmd := range commands {
		fmt.Printf("  %-8s  %s\n", cmd.Name, cmd.Description)
	}
}

func parseGlobal() (insecure, verbose bool, rest []string) {
	insecure = false
	verbose = false

	args := os.Args[1:]
	for len(args) > 0 {
		a := args[0]
		if !strings.HasPrefix(a, "-") {
			break
		}
		switch a {
		case "-k", "--insecure":
			insecure = true
		case "-v", "--verbose":
			verbose = true
		case "-h", "--help":
			showRootHelp()
			os.Exit(0)
		default:
			fmt.Fprintf(os.Stderr, "Unknown global option: %s\n", a)
			os.Exit(1)
		}
		args = args[1:]
	}
	return insecure, verbose, args
}

func mustBaseURL() string {
	u := os.Getenv("GFARM_HTTP_URL")
	if u == "" {
		fmt.Fprintln(os.Stderr, "Error: GFARM_HTTP_URL is required")
		os.Exit(1)
	}
	return strings.TrimSuffix(u, "/")
}

func main() {
	insecure, verbose, rest := parseGlobal()
	if len(rest) == 0 {
		showRootHelp()
		os.Exit(1)
	}
	command := rest[0]
	args := rest[1:]

	cfg := &Config{
		BaseURL:  mustBaseURL(),
		Verbose:  verbose,
		Insecure: insecure,
	}
	tr := &http.Transport{}
	if cfg.Insecure {
		tr.TLSClientConfig = &tls.Config{InsecureSkipVerify: true} //nolint:gosec
	}
	cfg.HTTPClient = &http.Client{Transport: tr}

	cmd, exists := commands[command]
	if !exists {
		fmt.Fprintf(os.Stderr, "Unknown command: %s\n", command)
		showRootHelp()
		os.Exit(1)
	}

	exitErr(cmd.Handler(cfg, args))
}
